force-app/main/default/classes/DMLOperations.cls

```diff
@@ -179,30 +281,55 @@ public with sharing class DMLOperations {
      * Example Contact.LastName = 'Jane' Link to Account.Name = 'Jane'
 	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
 	 */
-	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
-		// Loop through each Contact
-
-		// Extract the Account Name from Contact's LastName
-
-		// Get the account based on the name or create one if it doesn't exist
-
-		// Link the Contact to its Account using the Account Id
-
-		// Upsert the Contacts
-	}
+    public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
+
+        // Loop through each Contact in the provided list
+        for (Contact c : contactsToUpsert) {
+            // Check if the Contact's associated Account has a name
+            if (c.Account.Name != null) {
+                // If the Account's name is different from the Contact's last name, update the Account
+                if (!c.Account.Name.equals(c.LastName)) {
+                    Account newAccount = upsertAccount(c.LastName);
```

Comment:
Here you are doing DML inside a loop. It may work fine for small data volumes, but this can cause limit issues at scale. Can you refactor your code this to not do this?